Add diff mode support to git module

dagwieers opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task  -->

git
##### ANSIBLE VERSION

```
v2.0.2.0
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

This patch adds missing diff mode support to the git module.

It also fixes an issue related to check mode reporting a change when there was no change (run mode correctly reported no change). I added a remark below to the line where this was fixed.

dagwieers committed:
Add diffmode support to git module

This patch adds missing diffmode support to the git module.

dagwieers commented on source_control/git.py:
```diff
@@ -780,7 +804,10 @@ def main():
         if repo_updated is None:
             if module.check_mode:
-                module.exit_json(changed=True, before=before, after=remote_head)
+                result.update(changed=(result['before']!=remote_head), after=remote_head)
```
BUGFIX: This was required to ensure that check mode correctly returned a change compared to run mode. (Without this, check-mode reported a change where run-mode did not)

dagwieers commented:
@bcoca Why was it tagged `needs_revision` ?

bcoca commented:
sorry, thought your comment pointed out a needed change, but if it needs to be explained I would make it a comment in the code.

dagwieers commented:
It was a comment for the maintainer to point out a bug, so that it wouldn't be overlooked.

gregdek commented:
Thanks @dagwieers for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @dagwieers for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers commented:
@gregdek If the project would be merging stuff more often, we wouldn't have this many merge conflicts :-(

dagwieers commented:
ready_for_review

gregdek commented:
Thanks @dagwieers for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

robinro commented:
@dagwieers 
could you add an integration test for this feature? As far as I see there is no test yet for `--diff` at all, so this would be a valuable addition.
Maybe something along the line of 
- checkout commit A
- checkout commit B with --diff and grep for some part of the diff that should be there

dagwieers commented:
I wouldn't mind, however last time I looked into the unit-tests I got complete blocked because the tests would fail in awkward places (unrelated to my changes). And I simply gave up.

The unit-testing framework looks very complicated and that doesn't help in getting more unit-tests. I don't understand why we cannot simply have a directory where we can put playbooks, with asserts and they all get run one-by-one.

In my opinion that is much better than having to learn something new and complicated.

BTW Creating playbooks is exactly how I currently test and reproduce issues, so if you can reuse the stuff that people already write themselves. You solved 2 problems. And we can add every single minimal reproducer (that we often ask when people report issues) to another directory to ensure regressions don't return. CC @gregdek @robynbergeron

robinro commented:
@dagwieers 
The integration tests are mostly written by having playbooks run. In the `ansible/ansible` repo in`test/integration` there is a `role/test_git`. The chain things are run is:
- Makefile
- non_destructive.yml
- role/test_git/tasks/main.yml

If you want to test stdout output of ansible-playbook, you probably need to write a couple of shell lines in the Makefile yourself. I did this for example for ansible-pull.

The unit tests are somewhat more complicated, but I heard that there are plans to clean this up in the future.

robinro commented:
btw: I tried out your patch and it works well for me

dagwieers commented:
Nice to hear. I wouldn't have expected it otherwise ;-)

I will have to look at it again. The modification I made to the ordering of groups required me to write specific python with special libraries. Can't remember exactly.

Documentation regarding the unit tests is definitely welcome (and should probably be promoted as well).

dagwieers commented:
@robinro I am confused now that I wanted to look into it. The test-infrastructure is in the `ansible/ansible` repo. So how am I supposed to add tests if this has not been accepted yet ? The tests will only really work if the module is available.

robinro commented:
@dagwieers
It is really confusing and in my personal opinion not really ideal.
You need two pull requests for the two repos, one for your feature in core and one for the tests in ansible. If the feature gets merged first, the tests pass.
You should test the combination of course. If you can also test it on centos/... They have different (older) git version which could mess up things.
Travis runs the test/utils/run_tests.sh script, which runs docker containers for varios distros.

abadger commented on source_control/git.py:
```diff
@@ -368,6 +368,24 @@ def reset(git_path, module, dest):
+            result['diff'] = { 'prepared': '>> Failed to get proper diff between %s and %s:\n>> %s' % (result['before'], result['after'], err) }
+        else:
+            result['diff'] = { 'prepared': '>> Failed to get proper diff between %(before)s and %(after)s' % result }
+    return result
```
Programming style... Instead of operating directly on result, pass in before and after and return just the diff.  Then in calling code, do something like:

``` python
result['diff'] = get_diff(module, git_path, dest, result['before'], result['after'])
```

dagwieers replied:
Right, the problem is that I don't want to have an empty diff entry in my results when there is no diff information to expose. So that's why I did it like this. So in the end I still have to do the below for every occasion, and I don't think thatis necessarily better.

``` python
    diff = get_diff(module, git_path, dest, result['before'], result['after'])
    if diff:
        result['diff'] = diff
```

But I did it anyway :-)

abadger commented:
Logic looks good.  I made note of a change to make to the API of `get_diff()`.  Like @robinro I would like to see some integration tests to go with this ( add them to this role: https://github.com/ansible/ansible/blob/devel/test/integration/roles/test_git/tasks/main.yml )  (Utilize the tests by doing:

```
# Setup to run from the ansible checkout.  For instance:
# source ./hacking/env-setup
cd test/integration 
TEST_FLAGS='--tags test_git' make non_destructive
```

( Some tests require running as root as they change file ownership.  I don't believe the git tests do that)

However, after looking at this code, it's fairly self contained.  So tests would be nice to have but I won't block merging if they aren't submitted.

robinro commented:
@abadger the git test not only requires root, but is also destructive (removes /etc/ssh/ssh_known_hosts), see https://github.com/ansible/ansible-modules-core/issues/3281 and references there.

dagwieers committed:
Remodel get_diff() and calls to it

As proposed by @abadger

dagwieers commented:
ready_for_review

gregdek commented:
Thanks @dagwieers for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers committed:
Ensure we fetch the required object before performing a diff

Also we handle the return code ourselves, so don't leave this up to run_command().

gregdek commented:
@dagwieers A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@dagwieers Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers commented:
ready_for_review

gregdek commented:
Thanks @dagwieers for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @dagwieers for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@dagwieers A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@dagwieers Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@dagwieers A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

dagwieers commented:
Rebased, ready_for_review please.

This also fixes a bug in the git module we experience causing check_mode and run_mode to report differently.

gregdek commented:
Thanks @dagwieers for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

jctanner commented:
thanks @dagwieers !
